### ISSUE TITLE
Subtraction of mean on images that have white background

### ISSUE THREAD

**Issue by xsacha** [zisianw/FaceBoxes.PyTorch#56]
If I subtract the fixed mean (104,117,123) from the image, I get a much worse result on images that have a white background.
Is it better to subtract the actual mean of the image being tested?

**Comment by sfzhang15** [comment]
Empirically, we subtract the fixed mean (104,117,123) during training and testing should be consistent with this. It is strange that on images that have white background, subtracting the actual mean has better results than subtracting the fixed mean. Is this an individual case or a general phenomenon?

**Comment by xsacha** [comment]
It appears to be in general, images that have the face only and a blank colour background perform worse with the trained mean.

I found a suitable improvement is averaging the trained mean with the real mean and subtracting this instead. This improves results on all images including the tests.

**Comment by sfzhang15** [comment]
@xsacha 
Nice. Does this suitable for other images?

**Comment by xsacha** [comment]
Yes, I'm using this for all images:

C++
```C++
    static torch::Tensor trainedMean = torch::tensor({104.0f, 117.0f, 123.0f});
    auto imMean = tensorImage.view({-1, 3}).mean(0);
    tensorImage.sub_(torch::mean(torch::stack({imMean, trainedMean})));
```
Python
```Python
trained_mean = torch.tensor([104.0, 117.0, 123.0]).to(device)
...
    im_mean = img.view(-1, 3).mean(0)
    img -= torch.mean(torch.stack([im_mean, trained_mean]))
```

Basically averages the trained mean and the image mean and then subtracts that from the image.

**Comment by sfzhang15** [comment]
@xsacha 
Interesting operation. What is its improvement on AFW, PASCAL, FDDB?

**Comment by xsacha** [comment]
Quite small, about 0.15%. Didn't test FDDB.
It makes a much bigger difference on plain colour background images though.

**Comment by sfzhang15** [comment]
@xsacha 
Ok. Got it. Thank you.

**Comment by xsacha** [comment]
I retrained the model from my best checkpoint and found an improvement when using only the image mean rather than the fixed mean.

```Diff
-        image_t = _pad_to_square(image_t,self.rgb_means, pad_image_flag)
+        mean = cv2.mean(image_t)[:3]
+        image_t = _pad_to_square(image_t, mean, pad_image_flag)
         image_t, boxes_t = _mirror(image_t, boxes_t)
         height, width, _ = image_t.shape
-        image_t = _resize_subtract_mean(image_t, self.img_dim, self.rgb_means)
+        image_t = _resize_subtract_mean(image_t, self.img_dim, mean)
```
I got a new high accuracy with unmodified model (only pooling change) in AFW (99.12% AP, 99.98% precision @ 94% recall) and PASCAL (98.47% AP) as well as improvements for single colour background images.

The fixed mean you are using is from FG-Net I think? Any reason it is trained against it rather than image mean?

**Comment by sfzhang15** [comment]
@xsacha 
We follow SSD and Faster RCNN to subtract the fixed mean.
Subtraction of image mean seems to be a better choice for slightly higher performance while subtraction the fixed mean is simpler and doesn't require extra calculations.
